tests/Unit/Helpers/Evolution/DiscontinuousGalerkin/BoundaryConditions.hpp

```diff
@@ -0,0 +1,614 @@
+                  python_module, python_tag_function, face_mesh_velocity,
+                  interior_normal_covector, interior_face_and_volume_args...);
+            } catch (const std::exception& e) {
+              INFO("Failed python call with '" << e.what() << "'");
+              // Use REQUIRE(false) to print all the CAPTURE variables
+              REQUIRE(false);
+            }
+            CHECK_ITERABLE_CUSTOM_APPROX(
+                get<BoundaryCorrectionTag>(exterior_face_fields), python_result,
+                Approx::custom().epsilon(epsilon).scale(1.0));
+          });
+    };
+    // Since any of the variables in `exterior_face_fields` could be mutated
+    // from their projected state, we don't pass in the interior tags explicitly
+    // since that will likely give a false sense of "no aliasing"
+    apply_boundary_condition_impl(
+        apply_bc, interior_face_fields, bcondition_interior_tags{},
+        db::get<BoundaryConditionVolumeTags>(box_of_volume_data)...);
+  }
+}
+
+template <typename T, typename = std::void_t<>>
+struct get_boundary_conditions_impl {
+  using type = tmpl::list<>;
+};
+
+template <typename T>
+struct get_boundary_conditions_impl<
+    T, std::void_t<typename T::boundary_conditions_base>> {
+  using type = typename T::boundary_conditions_base::creatable_classes;
+};
+
+template <typename T>
+using get_boundary_conditions = typename get_boundary_conditions_impl<T>::type;
+}  // namespace detail
+
+/*!
+ * \brief Test a boundary condition against python code and that it satisfies
+ * the required interface.
+ *
+ * The specific boundary condition and system need to be explicitly given. Once
+ * the boundary condition and boundary correction base classes are
+ * listed/available in the `System` class we can remove needing to specify
+ * those. The `ConversionClassList` is forwarded to `pypp` to allow custom
+ * conversions of classes to python, such as analytic solutions or equations of
+ * state.
+ *
+ * - The random number generator is passed in so that the seed can be easily
+ *   controlled external. Use, e.g. `MAKE_GENERATOR(gen)` to create a generator.
+ * - The python function names are given in a `TaggedTuple`. A
+ *   `TestHelpers::evolution::dg::Tags::PythonFunctionForErrorMessage` tag must
+ *   be given for specifying the error message that the boundary condition
+ *   returns. In many cases this function will simply return `None`. The tag can
+ *   optionally specify the boundary correction for which it is to be used, so
+ *   that a different error message could be printed if a different boundary
+ *   correction is used.
+ *   The `TestHelpers::evolution::dg::Tags::PythonFunctionName` tag is used to
+ *   give the name of the python function for each return argument. The tags are
+ *   the input to the `package_data` function for Ghost boundary conditions, and
+ *   `::Tags::dt<evolved_var_tag>` for time derivative boundary conditions.
```

Comment:
I think it would be easier to add a short paragraph around line 499 where you give the big picture of what is compared and how.

Perhaps say something like (but with better English) "in general, a BC has several return arguments to set the BC values, and optionally returns an error string. therefore, the comparison implementation in python should have one function per return argument and one function for the error string." and also explain why it matters what the boundary correction is (if I understand correctly, this is because different corrections need different fields set by the BC?)

Then you can assume the reader has this context and simplify the description of the arguments here (it took me reading the comments and code several times to form an understanding of what was going on (and I'm not even sure I got it right!)).

Reply:
I'm adding the paragraph, but I'm not sure cutting here is helpful. Or at least, I'm not sure what details are extraneous. I find them helpful :shrug: